updated erc-4337 (#17)
eth-infinitism/ERCs: ERCS/erc-4337.md

| `paymasterVerificationGasLimit` | `uint256` | The amount of gas to allocate for the paymaster validation code (only if paymaster exists)        |
| `paymasterPostOpGasLimit`       | `uint256` | The amount of gas to allocate for the paymaster post-operation code (only if paymaster exists)    |
| `paymasterData`                 | `bytes`   | Data for paymaster (only if paymaster exists)                                                     |
| `signature`                     | `bytes`   | Data passed into the account to verify authorization                                              |

Users send `UserOperation` objects to a dedicated `UserOperation` mempool.

To prevent replay attacks (both cross-chain and multiple `EntryPoint` implementations), the `signature` should depend on `chainid` and the `EntryPoint` address.

Note that one [EIP-7702](./eip-7702.md) "authorization tuple" value can be provided alongside the `UserOperation` struct,
but "authorization tuples" are not included in the `UserOperation` itself.

### `EntryPoint` interface

> When passed on-chain the `EntryPoint` contract, the `Account` and the `Paymaster`, a "packed" version of the above structure called `PackedUserOperation` is used:

Reviewer: ```suggestion
When passed on-chain, to the `EntryPoint` contract, the `Account` and the `Paymaster`, a "packed" version of the above structure called `PackedUserOperation` is used:
```

}
```

The `userOpHash` is a hash over the `userOp` (except `signature`), `entryPoint` and `chainId`.

The account:

* MUST validate the caller is a trusted `EntryPoint`
* MUST validate that the signature is a valid signature of the `userOpHash`, and
  SHOULD return SIG_VALIDATION_FAILED (and not revert) on signature mismatch. Any other error MUST revert.
* SHOULD not return early when returning `SIG_VALIDATION_FAILED`. Instead, it SHOULD complete the normal flow to enable performing a gas estimation for the validation function.

> * MUST pay the `EntryPoint` (caller) at least the `missingAccountFunds` (which might be zero, in case the current account's deposit is high enough)

Reviewer: ```suggestion
* MUST pay the `EntryPoint` (caller) at least the `missingAccountFunds` (which might be zero, in case the current account's deposit is sufficient)
```

    );
```

### Support for [EIP-7702](./eip-7702.md) authorizations

> On networks with the [EIP-7702](./eip-7702.md) enabled, the `eth_sendUserOperation` method accepts an extra `eip7702Auth` parameter.

Reviewer: ```suggestion
On networks with [EIP-7702](./eip-7702.md) enabled, the `eth_sendUserOperation` method accepts an extra `eip7702Auth` parameter.
```

If this parameter is set, it should be a valid [EIP-7702](./eip-7702.md) authorization tuple, and signed by the `sender` address.

> The bundler MUST add all required `eip7702Auth` of all `UserOperations` in a bundle it to the `authorizationList` and execute

Reviewer: ```suggestion
The bundler MUST add all required `eip7702Auth` of all `UserOperations` in a bundle to the `authorizationList` and execute
```

> the bundle using a transaction of the `SET_CODE_TX_TYPE 0x04` transaction type.

Reviewer: ```suggestion
the bundle using a transaction type `SET_CODE_TX_TYPE`.
```

The factory has to be staked if it accesses global storage - see [reputation, throttling and banning section](#reputation-scoring-and-throttlingbanning-for-global-entities) for details.

NOTE: In order for the wallet to determine the "counterfactual" address of the wallet (prior to its creation),
it should make a static call to the `entryPoint.getSenderAddress()`

## Backwards Compatibility

This ERC does not change the consensus layer, so there are no backwards compatibility issues for Ethereum as a whole. Unfortunately it is not easily compatible with pre-[ERC-4337](./eip-4337.md) accounts, because those accounts do not have a `validateUserOp` function. If the account has a function for authorizing a trusted `UserOperation` submitter, then this could be fixed by creating an [ERC-4337](./eip-4337.md) compatible account that re-implements the verification logic as a wrapper and setting it to be the original account's trusted `UserOperation` submitter.

## Reference Implementation

See `https://github.com/eth-infinitism/account-abstraction/tree/main/contracts`

## Security Considerations

> The `EntryPoint` contract will need to be very heavily audited and formally verified, because it will serve as a central trust point for _all_ [ERC-4337]. In total, this architecture reduces auditing and formal verification load for the ecosystem, because the amount of work that individual _accounts_ have to do becomes much smaller (they need only verify the `validateUserOp` function and its "check signature and pay fees" logic) and check that other functions are `msg.sender == ENTRY_POINT` gated (perhaps also allowing `msg.sender == self`), but it is nevertheless the case that this is done precisely by concentrating security risk in the `EntryPoint` contract that needs to be verified to be very robust.

Reviewer: ```suggestion
The `EntryPoint` contract will need to be audited and formally verified, because it will serve as a central trust point for _all_ [ERC-4337]. In total, this architecture reduces auditing and formal verification load for the ecosystem, because the amount of work that individual _accounts_ have to do becomes much smaller (they need only verify the `validateUserOp` function and its "check signature and pay fees" logic) and check that other functions are `msg.sender == ENTRY_POINT` gated (perhaps also allowing `msg.sender == self`), but it is nevertheless the case that this is done precisely by concentrating security risk in the `EntryPoint` contract that needs to be verified to be very robust.
```
